File: src/main/java/de/bixilon/minosoft/data/entities/VillagerData.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.entities;

import de.bixilon.minosoft.data.MapSet;
import de.bixilon.minosoft.data.VersionValueMap;

public record VillagerData(VillagerTypes type, VillagerProfessions profession, VillagerLevels level) {

    public VillagerData(int type, int profession, int level, int versionId) {
        this(VillagerTypes.byId(type), VillagerProfessions.byId(profession, versionId), VillagerLevels.byId(level));
    }

    public enum VillagerProfessions {
        NONE(new MapSet[]{new MapSet<>(451, 0)}),
        ARMORER(new MapSet[]{new MapSet<>(451, 1)}),
        BUTCHER(new MapSet[]{new MapSet<>(0, 4), new MapSet<>(451, 2)}),
        CARTOGRAPHER(new MapSet[]{new MapSet<>(451, 3)}),
        CLERIC(new MapSet[]{new MapSet<>(451, 4)}),
        FARMER(new MapSet[]{new MapSet<>(0, 0), new MapSet<>(451, 5)}),
        FISHERMAN(new MapSet[]{new MapSet<>(451, 6)}),
        FLETCHER(new MapSet[]{new MapSet<>(451, 7)}),
        LEATHER_WORKER(new MapSet[]{new MapSet<>(451, 8)}),
        LIBRARIAN(new MapSet[]{new MapSet<>(0, 1), new MapSet<>(451, 9)}),
        MASON(new MapSet[]{new MapSet<>(451, 10)}),
        NITWIT(new MapSet[]{new MapSet<>(0, 5), new MapSet<>(451, 11)}),
        SHEPHERD(new MapSet[]{new MapSet<>(451, 12)}),
        TOOL_SMITH(new MapSet[]{new MapSet<>(451, 13)}),
        WEAPON_SMITH(new MapSet[]{new MapSet<>(451, 14)}),
        PRIEST(new MapSet[]{new MapSet<>(0, 2), new MapSet<>(451, -1)}),
        BLACKSMITH(new MapSet[]{new MapSet<>(0, 3), new MapSet<>(451, -1)}),

        HUSK(new MapSet[]{new MapSet<>(204, 5), new MapSet<>(315, -100)}), // ToDo
        ZOMBIE(new MapSet[]{new MapSet<>(204, -1), new MapSet<>(315, -100)}); // used earlier (ZombieVillagerMeta)

        final VersionValueMap<Integer> valueMap;

        VillagerProfessions(MapSet<Integer, Integer>[] values) {
            valueMap = new VersionValueMap<>(values, true);
        }

        public static VillagerProfessions byId(int id, int versionId) {
            for (VillagerProfessions profession : values()) {
                if (profession.getId(versionId) == id) {
                    return profession;
                }
            }
            return null;
        }

        public int getId(int versionId) {
            return valueMap.get(versionId);
        }
    }

    public enum VillagerTypes {
        DESSERT,
        JUNGLE,
        PLAINS,
        SAVANNA,
        SNOW,
        SWAMP,
        TAIGA;

        public static VillagerTypes byId(int id) {
            return values()[id];
        }
    }

    public enum VillagerLevels {
        NOVICE,
        APPRENTICE,
        JOURNEYMAN,
        EXPERT,
        MASTER;

        public static VillagerLevels byId(int id) {
            return values()[id];
        }
    }
}

File: src/main/java/de/bixilon/minosoft/data/entities/VillagerData.kt
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */
package de.bixilon.minosoft.data.entities

import de.bixilon.minosoft.data.VersionValueMap
import de.bixilon.minosoft.protocol.protocol.ProtocolVersions.*

data class VillagerData(val type: VillagerTypes, val profession: VillagerProfessions, val level: VillagerLevels) {
    constructor(type: Int, profession: Int, level: Int, versionId: Int) : this(VillagerTypes.values()[type], VillagerProfessions.byId(profession, versionId)!!, VillagerLevels.values()[level])

    enum class VillagerProfessions(val values: Map<Int, Int>) {
        NONE(mapOf(V_18W50A to 0)),
        ARMORER(mapOf(V_18W50A to 1)),
        BUTCHER(mapOf(LOWEST_VERSION_SUPPORTED to 4, V_18W50A to 2)),
        CARTOGRAPHER(mapOf(V_18W50A to 3)),
        CLERIC(mapOf(V_18W50A to 4)),
        FARMER(mapOf(LOWEST_VERSION_SUPPORTED to 0, V_18W50A to 5)),
        FISHERMAN(mapOf(V_18W50A to 6)),
        FLETCHER(mapOf(V_18W50A to 7)),
        LEATHER_WORKER(mapOf(V_18W50A to 8)),
        LIBRARIAN(mapOf(LOWEST_VERSION_SUPPORTED to 1, V_18W50A to 9)),
        MASON(mapOf(V_18W50A to 10)),
        NITWIT(mapOf(LOWEST_VERSION_SUPPORTED to 5, V_18W50A to 11)),
        SHEPHERD(mapOf(V_18W50A to 12)),
        TOOL_SMITH(mapOf(V_18W50A to 13)),
        WEAPON_SMITH(mapOf(V_18W50A to 14)),
        PRIEST(mapOf(LOWEST_VERSION_SUPPORTED to 2, V_18W50A to -1)),
        BLACKSMITH(mapOf(LOWEST_VERSION_SUPPORTED to 3, V_18W50A to -1)),
        HUSK(mapOf(V_1_10_PRE1 to 5, V_1_11 to -100)),  // ToDo
        ZOMBIE(mapOf(V_1_10_PRE1 to -1, V_1_11 to -100));

        // used earlier (ZombieVillagerMeta)
        private val valueMap = VersionValueMap(values)


        fun getId(versionId: Int): Int {
            return valueMap[versionId]
        }

        companion object {
            @JvmStatic
            fun byId(id: Int, versionId: Int): VillagerProfessions? {
                for (profession in values()) {
                    if (profession.getId(versionId) == id) {
                        return profession
                    }
                }
                return null
            }
        }

    }

    enum class VillagerTypes {
        DESSERT, JUNGLE, PLAINS, SAVANNA, SNOW, SWAMP, TAIGA;

        companion object {
            private val VILLAGER_TYPES: Array<VillagerTypes> = values()

            @JvmStatic
            fun byId(id: Int): VillagerTypes {
                return VILLAGER_TYPES[id]
            }
        }
    }

    enum class VillagerLevels {
        NOVICE, APPRENTICE, JOURNEYMAN, EXPERT, MASTER;

        companion object {
            private val VILLAGER_LEVELS: Array<VillagerLevels> = values()

            @JvmStatic
            fun byId(id: Int): VillagerLevels {
                return VILLAGER_LEVELS[id]
            }
        }
    }
}
